src/aleph/http.clj

```diff
@@ -358,96 +362,104 @@
+
+                                                   ;; request timeout triggered
+                                                   (d/catch' TimeoutException
+                                                             (fn [^Throwable e]
+                                                               (d/error-deferred (RequestTimeoutException. e))))
+
+                                                   ;; clean up the connection
+                                                   (d/chain'
+                                                    (fn cleanup-conn [rsp]
+
+                                                      ;; either destroy/dispose of the conn, or release it back for reuse
+                                                      (-> (:aleph/destroy-conn? rsp)
+                                                          (maybe-timeout! read-timeout)
+
+                                                          ;; read timeout triggered
+                                                          (d/catch' TimeoutException
+                                                                    (fn [^Throwable e]
+                                                                      (log/trace "Request timed out.")
+                                                                      (d/error-deferred (ReadTimeoutException. e))))
+
+                                                          (d/chain'
+                                                           (fn [early?]
+                                                             (if (or early?
+                                                                     (not (:aleph/keep-alive? rsp))
+                                                                     (<= 400 (:status rsp)))
+                                                               (do
+                                                                 (log/trace "Connection finished. Disposing...")
+                                                                 (flow/dispose pool k conn))
+                                                               (flow/release pool k conn)))))
+                                                      (-> rsp
+                                                          (dissoc :aleph/destroy-conn?)
+                                                          (assoc :connection-time (- end start)))))))))
+
+                                         (fn handle-response [rsp]
+                                           (->> rsp
+                                                (middleware/handle-cookies req)
+                                                (middleware/handle-redirects request req))))))))))))
+                      req))]
+      (d/connect response result)
+      (d/catch' result
+                (fn [e]
+                  (log/trace e "Request failed. Disposing of connection...")
+                  (@dispose-conn!)
+                  (d/error-deferred e)))
```

Comment:
Won't it log "request failed" even if the _user_ puts it in an error state? That'll be confusing.

Maybe "Request failed or canceled."?

Reply:
Fair enough, Or we could log a different message depending on `e`'s type?

BTW I just noticed that I was somewhat wrong about the original error handlers being redundant: the original handlers used `log/error` while this one here uses `log/trace`, so there's a user-visible change. I for one would like to change it this way, though, as I feel the `log/error` doesn't add much for users which I expect to deref the response anyway, in which case they will receive the exception which would then (usually) generate a corresponding error log. And if the user decides to handle e.g. a timeout error explicitly, it would also be nicer to not spam the error log. WDYT?

Reply:
Oh yeah, also note that the exception's message will say "Request cancelled" in case the user cancelled the request.

Reply:
(and since that's logged right along with the "Request failed ..." it might be alright as-is after all?)